Import Crosslink design & analysis from hackmd pages. (#138)
Electric-Coin-Company/tfl-book: src/design/goals.md

# Design Goals

Here we strive to lay out our high level TFL design goals.

## Goals, Design, and Trade-offs

Here we lay out _ideal_ goals. As we develop a complete design, we are likely to inevitably encounter trade-offs some of which may preclude achieving the full idealized goals. Wherever possible, we motivate design decisions by these goals, and when goals are impacted by trade-offs we describe that impact and the rationale for the trade-off decision.

> For example, one ideal user experience goal below is to avoid disruption to existing wallets. However, the [Crosslink](../terminology.md#definition-crosslink) construction may require wallets to to alter their context of valid transactions differently from the current [NU5](../terminology.md#definition-nu5) protocol.

Reviewer: ```suggestion
For example, one ideal user experience goal below is to avoid disruption to existing wallets. However, the [Crosslink](../terminology.md#definition-crosslink) construction may require wallets to alter their context of valid transactions differently from the current [NU5](../terminology.md#definition-nu5) protocol.
```

- All currently supported wallet user experience should continue to operate seamlessly without change during or after protocol transitions. This covers the use of addresses, payment flow, transfers, ZEC supply cap and issuance rate, backup/restore, and other features users currently rely on.
- There must be no security or safety degradation due to wallet user behavior introduced by PoS transitions, assuming users follow their current behaviors unchanged and continue to use the same cognitive model of the impacts of their behaviors. This goal encompasses all of security and safety, including privacy and transparency or more explicit disclosures.
- The protocol should enable users of shielded mobile wallets to delegate ZEC to PoS consensus providers and earn a return on that ZEC coming via ZEC issuance or fees. Doing this may expose users to a risk of loss of delegated ZEC (such as through “slashing fees”). The protocol must guarantee that PoS consensus providers have no discretionary control over such delegated funds (including that they cannot steal those funds).

> - For any hybrid PoW/PoS protocol (including the PoW+TFL protocol we’re proposing), the process and UX of mining remains unchanged except that the return on investment may be altered. This is true both of consensus level block miners (ie mining pools and solo miners) and mining pool participants.

Reviewer: ```suggestion
- For any hybrid PoW/PoS protocol (including the PoW+TFL protocol we’re proposing), the process and UX of mining remains unchanged except that the return on investment may be altered. This is true both of consensus-level block miners (i.e. mining pools and solo miners) and mining pool participants.
```

> - Any hybrid PoW/PoS protocol (including PoW+TFL) block explorers will continue to function with the same UX through transitions in-so-far as displaying information about transactions, the mempool, and blocks.

Reviewer: ```suggestion
- Any hybrid PoW/PoS protocol (including PoW+TFL) block explorers will continue to function with the same UX through transitions as far as displaying information about transactions, the mempool, and blocks.
```

- Block explorers and other network metrics sites may require UX changes with respect to mining rewards and issuance calculations.
- Network metrics sites may require UX changes with respect to the p2p protocol or other network-specific information.
- Users can rely on [assured finality](../terminology.md#definition-assured-finality) with an expected time-to-finality of <30m.[^req-ttf-30m]

## Developer Experience Goals

For a full PoS transition, ecosystem developers for products such as consensus nodes, wallets, mining services, chain analytics, and more will certainly need to update their code to support transitions. However, we carve out a few goals as an exception to this for this category of users:

> - Wallet developers should not be required to make any changes through protocol transitions as long as they rely solely on the lightwalletd protocol or a full node API (such as the zcashd RPC interface).

Reviewer: I think this is an overspecification.

Author: Let's address it in future work after we discuss it.

I'm happy to recategorize this as a desideratum. The big picture is "It would be really nice if wallets had minimal upgrade maintenance burden." -but I agree this is overspecified.

For one thing, what if lightwalletd is replaced prior to TFL activation? Maybe it could be worded more generally, something about light wallet support being localized to a lightwallet protocol / SDKs such that wallet apps themselves are minimally impacted? Hrm, that still seems too prescriptive. We could just use my quote above in this comment as a very general desideratum.
